### DISCUSSION TITLE
[BP#71] Utiliser des variables statiques

### DISCUSSION THREAD

**Discussion by devcrafting** [cnumr/best-practices#85] [Bonnes pratiques] [answered]
J'ai du mal à comprendre cette BP, elle me semble pousser une pratique de qualité logicielle assez discutable : le statique est une des plus grosses problématiques de qualité à mon sens.

A noter que des plateformes avec mémoire partagée entre les requêtes permettent de faire de la gestion de cycle de vie des instances comme un DataParser partagé par toutes les requêtes via inversion de dépendances (le D de SOLID + éventuellement injection de dépendances) => à pousser particulièrement pour les services sans état

Bon après je reste circonspect sur l'impact sobriété réel de cela : comme d'autres BP, dans des bases de code avec beaucoup de soucis (notamment de qualité), ça peut être important mais sinon c'est probablement pas grand chose.

**Comment by ACTLEM** [comment]
Dans la BP, il est stipulé que c'est efficace surtout pour des traitements importants. Après, cela reste spécifique PHP, je ne sais pas si c'est encore vraiment d'actualité sur les différences en termes d'empreinte mémoire et consommation CPU. A creuser. Si c'est le cas, il manque peut-être une information sur le fait que c'est uniquement valable en PHP car comme tu le dis, il y a l'inversion de dépendances pour gérer cela.

**Reply by tbroyer** [reply]
Ça a peu à voir avec l'injection de dépendance, et ça n'a rien de spécifique à PHP.

Si j'utilise de l'injection de dépendance mais que chaque objet est reconstruit à chaque fois (pas de singleton), alors un objet "gourmand" à construire aura le même effet.
L'idée ici est donc bien générale et générique, à penser au cycle de vie de ses objets et faire en sorte de réutiliser les objets "coûteux" à construire/initialiser, et donc ne pas exclure les singletons (mais on pourra aussi avoir des durées de vie plus courtes que le processus, tout en réutilisant les objets ; des "pseudos-singletons" qu'on s'autorise à détruire au bout d'un certain temps de non-utilisation, ou liés à un ensemble "d'écrans" –plutôt pour le frontend donc– et qu'on détruira si on passe à autre chose, ou encore même si c'est une approche qui ne me plaît pas, des objets stockés dans la session de l'utilisateur côté serveur, avec donc une durée de vie limitée dans le temps mais tout de même bien plus longue qu'une seule requête/réponse)

**Reply by devcrafting** [reply]
J'ai bien parlé de gestion de cycle de vie des objets et d'inversion de dépendance (à ne pas confondre avec injection de dépendance, même si lié). L'inversion de dépendance permet d'externaliser le cycle de vie d'une dépendance, et en gros à chaque fois qu'on a la tentation d'utiliser une dépendance statique, le réflexe devrait être d'inverser la dépendance.

Concernant les singletons, c'est une stratégie de cycle de vie qui implémentée en statique ne me semble pas intéressante, car elle crée un couplage, le "client" du singleton sait explicitement que c'est un singleton alors que ce n'est pas desa responsabilité. A noter qu'avec l'inversion de dépendance, on peut alors injecter des dépendances dont le cycle de vie est géré indépendamment, avec différentes stratégies/scope: singleton, par requête, par session...

**Comment by arnolem** [comment]
Est-ce que l'on ne serait pas, encore une fois, dans un sujet très anecdotique dans la majorité des cas ?
Il faudrait prouver l'effet de cette mesure pour voir si ça rentre vraiment dans un TOP 100 des pratiques d'éco-conception.

**Reply by devcrafting** [reply]
Je me demande en effet si le sujet de construction/destruction d'objet est vraiment un sujet. C'est un serpent de mer de l'optimisation prématurée: il faut faire attention à ne pas faire n'importe quoi (et s'appuyer sur des principes de qualité logicielle comme SOLID pour éviter de tomber dans le piège trop vite), mais on n'optimise que si on mesure clairement le problème.

Bon après je me dis que mentionner cette pratique permet de mettre en avant des pratiques de qualité logicielle qui limite l'apparition de ce genre de problème.

**Comment by ACTLEM** [comment, marked as answer]
Note équipe projet: Pour la suppression de cette bonne pratique qui reste anecdotique d'un point de vue écoconception.

Issue créée: https://github.com/cnumr/best-practices/issues/167
